### PR TITLE
kubeadm: Enable the Node Authorizer/Admission plugin in v1.7

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

This is similar to https://github.com/kubernetes/kubernetes/pull/46796, but for kubeadm.
Basically it was a part of https://github.com/kubernetes/kubernetes/pull/46796, but there were some other upgradability and compability concerns for kubeadm I took care of while working today.

Example:

```console
$ kubeadm init --kubernetes-version v1.7.0-beta.0
[kubeadm] WARNING: kubeadm is in beta, please do not use it for production clusters.
[init] Using Kubernetes version: v1.7.0-beta.0
[init] Using Authorization mode: [RBAC Node]
...
$ sudo kubectl --kubeconfig=/etc/kubernetes/kubelet.conf get secret foo
Error from server (Forbidden): User "system:node:thegopher" cannot get secrets in the namespace "default".: "no path found to object" (get secrets foo)

$ echo '{"apiVersion":"v1","kind":"Node","metadata":{"name":"foo"}}' | sudo kubectl create -f - --kubeconfig=/etc/kubernetes/kubelet.conf
Error from server (Forbidden): error when creating "STDIN": nodes "foo" is forbidden: node thegopher cannot modify node foo
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

Depends on https://github.com/kubernetes/kubernetes/pull/46864 (uses that PR as a base, will rebase once it's merged)

Please only review the second commit. Will also fix tests in a minute.

**Release note**:

```release-note
kubeadm: Enable the Node Authorizer/Admission plugin in v1.7 
```
@mikedanese @liggitt @pipejakob @roberthbailey @jbeda @timothysc

### CODE REVIEW

File: cmd/kubeadm/app/cmd/defaults.go [hunk 1]
```diff
@@ -53,6 +53,9 @@ func setInitDynamicDefaults(cfg *kubeadmapi.MasterConfiguration) error {
 	if k8sVersion.LessThan(kubeadmconstants.MinimumControlPlaneVersion) {
 		return fmt.Errorf("this version of kubeadm only supports deploying clusters with the control plane version >= %s. Current version: %s", kubeadmconstants.MinimumControlPlaneVersion.String(), cfg.KubernetesVersion)
 	}
+	if k8sVersion.AtLeast(kubeadmconstants.MinimumNodeAuthorizerVersion) {
+		cfg.AuthorizationModes = append(cfg.AuthorizationModes, "Node")
```
Comment (liggitt) [inline cmd/kubeadm/app/cmd/defaults.go]: I would put this at the beginning, it's likely to be the fastest one, and we'd want to see the denial messages in the log if it was preventing a node from doing something

File: cmd/kubeadm/app/master/manifests.go [hunk 2]
```diff
@@ -43,6 +43,9 @@ import (
 const (
 	DefaultCloudConfigPath = "/etc/kubernetes/cloud-config"
 
+	defaultv16AdmissionControl = "NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota,DefaultTolerationSeconds"
```
Comment (liggitt) [inline cmd/kubeadm/app/master/manifests.go]: ResourceQuota should always be last

File: cmd/kubeadm/app/phases/apiconfig/clusterroles.go [hunk 3]
```diff
@@ -84,6 +85,10 @@ func CreateRBACRules(clientset *clientset.Clientset) error {
 	if err := createClusterRoleBindings(clientset); err != nil {
 		return err
 	}
+	if err := deletePermissiveNodesBindingWhenUsingNodeAuthorization(clientset); err != nil {
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/apiconfig/clusterroles.go]: I would do this separately from the CreateRBACRules step... logically, it should only happen when both the Node and RBAC authorizers are in place, and doesn't the step that calls CreateRBACRules() know the version of the master it is setting up? I wouldn't expect to need to do discovery here

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 4]
```diff
@@ -58,6 +59,9 @@ func CreateInitKubeConfigFiles(masterEndpoint, pkiDir, outDir string) error {
 	if err != nil {
 		return err
 	}
+	// Lowercasing here is needed, because Node names in kubernetes are lower-case, and the Node Admission plugin strips "system:node:" from the CN below
+	// and compares it strictly to the (kubernetes lower-cased) Node name.
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: point to the `nodeutil.GetHostname` kubelet method this is mimicking

Comment (liggitt) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: also note that this will need to change if cloud providers ever get involved

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 5]
```diff
@@ -58,6 +59,9 @@ func CreateInitKubeConfigFiles(masterEndpoint, pkiDir, outDir string) error {
...
+	hostname := strings.ToLower(hostname)
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: and call this `masterNodeName`

Comment (liggitt) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: currently, kubeadm skips the cert generation step if the cert already exists. if kubeadm wants to support upgrades, it will now need to read an existing client cert and verify it is valid (not expired and subject matches what it has been asked to generate), and regenerate the cert if needed

File: cmd/kubeadm/app/master/manifests.go (hunk 2 above)
Comment (luxas) [inline cmd/kubeadm/app/master/manifests.go]: Haven't actually touched this, I guess this came from the "add DefaultTolerationsSeconds everywhere" PR or so. Will fix.

File: cmd/kubeadm/app/cmd/defaults.go (hunk 1 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/defaults.go]: Great, will do.

**Comment by luxas** [comment]
@liggitt Updated to address your feedback. The tests are expected to fail, but it would be great if I could get a soft LGTM for the second commit and then fixup tests after the rebase...

File: cmd/kubeadm/app/phases/apiconfig/clusterroles.go [hunk 6]
```diff
@@ -84,6 +85,10 @@ func CreateRBACRules(clientset *clientset.Clientset) error {
...
+		// This is not a fatal error, proceed as usual but log it.
+		fmt.Println("[apiconfig] Failed to remove the permissive 'system:node' ClusterRoleBinding: %v", err)
```
Comment (pipejakob) [inline cmd/kubeadm/app/phases/apiconfig/clusterroles.go]: If it's not a fatal error, I would expect more explanation in the message itself about it being innocuous (unless we want to get a lot of issues from people asking about it). Also, is there really no case where this could fail and we would care? Should it be purely best-effort?

I would imagine that if someone upgrades their kubeadm cluster to 1.7, and sees the new authorizer/admission plugin in place, and reads about the new security features, it should be safe for them to assume they're actually more secure (and not that a random transient failure might have actually caused this less-secure rolebinding to persist during the upgrade). I would understand ignoring NotFound errors where there is nothing to do, but to ignore all errors and succeed the operation feels wrong to me.

**Review by pipejakob — COMMENTED** [review]
Based on my reading of https://github.com/kubernetes/kubernetes/pull/46796, I'd give your second commit a soft LGTM. But of course, more (and updated) tests!

File: cmd/kubeadm/app/phases/apiconfig/clusterroles.go (hunk 6 above)
Comment (liggitt) [inline cmd/kubeadm/app/phases/apiconfig/clusterroles.go]: Agree, notfound is ignorable

File: cmd/kubeadm/app/phases/apiconfig/clusterroles.go (hunk 3 above)
Comment (pipejakob) [inline cmd/kubeadm/app/phases/apiconfig/clusterroles.go]: To @luxas's point, I think having this step in `CreateRBACRules` is more palatable now that we're making all of these steps idempotent, so it's really `CreateOrUpdateRBACRules` or `MakeSureAllRBACRulesAreCorrect`. This will likely be externalized to a command like `kubeadm phases rbac`, which means "do everything you need to in order to make sure RBAC roles are set up correctly," which involves potentially deleting outdated roles.

Comment (liggitt) [inline cmd/kubeadm/app/phases/apiconfig/clusterroles.go]: Then the target version being deployed should be part of the input. I'd point out that tightening roles is generally something you would do *after* a deployment completed, because of version skew between components

Comment (pipejakob) [inline cmd/kubeadm/app/phases/apiconfig/clusterroles.go]: That's fair. It's also worthwhile waiting until we have the phase subcommand fully spec'd out (with interested consumers) before trying to implement toward it.

File: cmd/kubeadm/app/master/manifests.go [hunk 7]
```diff
@@ -42,6 +43,9 @@ import (
 const (
 	DefaultCloudConfigPath = "/etc/kubernetes/cloud-config"
 
+	defaultv16AdmissionControl = "NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,ResourceQuota"
+	defaultv17AdmissionControl = "NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,ResourceQuota"
```
Comment (timothysc) [inline cmd/kubeadm/app/master/manifests.go]: Couldn't we check the version and add NodeRestriction?

**Review by timothysc — COMMENTED** [review]
minor comments.

File: cmd/kubeadm/app/master/manifests.go (hunk 7 above)
Comment (luxas) [inline cmd/kubeadm/app/master/manifests.go]: That's basically what this is doing, but since it should be in between DefaultTolerationSeconds and ResourceQuota I felt like this was the most basic and stable way to do it. Didn't want to get too fancy with slices and such things although that would also work

**Comment by dchen1107** [comment]
I added /approved label since the pr does have the associated issue. But the pr is not lgtmed before the code freeze. I am leaving the @kubernetes/sig-cluster-lifecycle-misc to decide if pr is for fixing a regression bug or a new feature. If it is for new feature, there should be an exception request.

cc/ @kubernetes/kubernetes-release-managers

**Comment by luxas** [comment]
Rebased upon https://github.com/kubernetes/kubernetes/pull/46864, so ready to be merged BUT this will have to wait until https://github.com/kubernetes/kubernetes/pull/47021 is in.

@liggitt Updated to address your feedback, PTAL

@k8s-bot pull-kubernetes-e2e-kubeadm-gce test this

I think we'll have a green run now... unless we run into quota issues for the job

**Comment by timothysc** [comment]
@dchen1107 this is more an enablement/fix around a feature that was added to this release.  Let me know if think we need an exception on this. 

cc/ @kubernetes/kubernetes-release-managers

**Comment by luxas** [comment]
@dchen1107 So this PR was initially a commit in #46796. #46796 was approved/lgtm'd before the code freeze. Then I noticed a potential source to a problem in the commit that enabled the Node Authorizer for kubeadm in PR #46796, so I asked if I could move the enablement of the Node Authorizer for kubeadm to an other PR instead. So I did. That is this PR.

It should also be mentioned that the Node Authorizer feature merged one-two days before the code freeze, leaving us (kubeadm folks) with a minimal amount of time to enable it in the code. This change is pretty straightforward.

This PR also fixes an other breakage made in the massive code freeze rush. The `Initializers` admission controller was (blindly) added in an automated fashon across the codebase. However, that makes it impossible to deploy v1.6 clusters with kubeadm if it insists to _always_ set the Initializers admission controller (which didn't exist in v1.6).

TL;DR; This is a bug fix for kubeadm AND an enablement of a feature that did make the deadline (the Node Authorizer feature), where the original enablement PR also was lgtm'd before the code freeze.

File: cmd/kubeadm/app/cmd/defaults_test.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		authzModes []string
+		version    string
+		expected   []string
+	}{
+		{[]string{"RBAC"}, "v1.6.0", []string{"RBAC"}},
+		{[]string{"RBAC", "ABAC"}, "v1.6.4", []string{"RBAC", "ABAC"}},
+		{[]string{"RBAC"}, "v1.7.0", []string{"Node", "RBAC"}},
+		{[]string{"RBAC", "Webhook"}, "v1.7.0", []string{"Node", "RBAC", "Webhook"}},
+		{[]string{"RBAC", "Webhook", "Node"}, "v1.7.0", []string{"RBAC", "Webhook", "Node"}},
+		{[]string{"Node", "RBAC", "Webhook"}, "v1.7.0", []string{"Node", "RBAC", "Webhook"}},
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/defaults_test.go]: Could you add a -beta example test too for completeness.

**Review by timothysc — COMMENTED** [review]
Only a minor comment.

**Comment by dchen1107** [comment]
Re: https://github.com/kubernetes/kubernetes/pull/46879#issuecomment-306607923

Thanks for the detail information. I am ok with this.  @kubernetes/kubernetes-release-managers WDYT?

**Comment by luxas** [comment]
Thank you @dchen1107!

@timothysc Updated the unit test to test beta versions as well

I'm feature-gating the Node Authorizer and Initializer admission controller behind beta.1, since some required parts of that work was done after beta.0.

So after the beta.1 cut, this PR may be merged and the kubeadm e2e CI will turn green again @pipejakob :)

**Comment by timothysc** [comment]
@luxas @liggitt Are there any e2e's that validate this feature?

**Comment by luxas** [comment]
@timothysc I was thinking we could start with adding the two lines in the PR description as tests in kubernetes-anywhere. It would be super cool if there were ready-made e2e's as well (which there probably is). In case there is, we should match them in the kubeadm job for sure

**Comment by liggitt** [comment]
When the confirmance suite is run with this configuration on, every time a node runs a pod and reports pod status, this authorizer and admission plugin are exercised.

Negative verification exists as an integration test only today - https://github.com/kubernetes/kubernetes/blob/master/test/integration/auth/node_test.go

**Comment by timothysc** [comment]
> "with this configuration on" 

What is the plan for defaulting on?

**Comment by liggitt** [comment]
> What is the plan for defaulting on?

Up to each deployment method. gce/gke is enabling in https://github.com/kubernetes/kubernetes/pull/46796

with this PR, the conformance e2e suite run against a kubeadm-setup cluster would exercise the authorizer/admission

**Comment by liggitt** [comment]
don't the bugfixes need to be made now, separately, in order for beta.1 to cut if we want the kubeadm CI job to be release blocking?

**Comment by luxas** [comment]
@liggitt No, actually not required. The kubeadm job builds kubeadm from master but uses control plane images from latest release. So once beta.1 is cut with Initializers support and this PR is merged it should go green.

I'm open to merging this as-is now though if we want to have this in beta.1, either way works

**Comment by luxas** [comment]
Removing do-not-merge as beta.1 was cut :+1: 

@k8s-bot pull-kubernetes-e2e-kubeadm-gce test this

/retest
